packages/financial-templates-lib/src/clients/InsuredBridgeL2Client.ts

```diff
@@ -122,14 +128,88 @@ export class InsuredBridgeL2Client {
+      );
+    }
+
+    // All events were found in all providers, can return any of the event data arrays
+    return eventsData.events[0];
   }
 
-  async getWhitelistTokenEvents(blockSearchConfig: { fromBlock: number; toBlock: number }): Promise<EventData[]> {
-    return await this.bridgeDepositBox.getPastEvents("WhitelistToken", blockSearchConfig);
+  async getWhitelistTokenEvents(blockSearchConfig: BlockSearchConfig): Promise<EventData[]> {
+    const eventsData = await this.getEventsForMultipleProviders(
+      [this.l2Web3].concat(this.fallbackL2Web3s), // TODO: This might create redundant web3 calls for the same provider
+      this.bridgeDepositBox.options.jsonInterface,
+      this.bridgeDepositAddress,
+      "WhitelistToken",
+      blockSearchConfig
+    );
+    if (!eventsData.success) {
+      // TODO: Do something with eventsData.missingEvents array.
+      throw new Error(
+        `${eventsData.missingEvents.length} WhitelistToken events found in fallback l2 provider not found in all l2 web3 providers`
+      );
+    }
+
+    // All events were found in all providers, can return any of the event data arrays
+    return eventsData.events[0];
   }
 
+  /**
+   * Fetches specified contract event data for all input web3 providers. Returns false if any of the events found with
+   * one provider are NOT matched exactly in ALL of the other providers' event arrays.
+   * @param web3s Web3 providers to check for target event.
+   * @param contractAbi Contract ABI to query target event on.
+   * @param contractAddress Contract address to query target event on.
+   * @param eventName Name of target event.
+   * @param blockSearchConfig Target event search config
+   * @returns Object containing success of event query, missing events not found in all providers, and all event data
+   */
+  async getEventsForMultipleProviders(
+    web3s: Web3[],
+    contractAbi: any[],
+    contractAddress: string,
+    eventName: string,
+    blockSearchConfig: BlockSearchConfig
+  ): Promise<{ success: boolean; missingEvents: EventData[]; events: EventData[][] }> {
+    const events = await Promise.all(
+      web3s.map((_web3) => {
+        const _contract = new _web3.eth.Contract(contractAbi, contractAddress);
+        return _contract.getPastEvents(eventName, blockSearchConfig);
+      })
+    );
+
+    // Create array of data uniquely identifying each event that we'll use to compare across providers.
+    const uniqueEventKeys = events[0].map((event) => event.transactionHash);
+
+    // Compare unique event keys from all providers with the provider at index 0, and keep track of mismatching events.
+    const missingEvents: EventData[] = [];
+    for (let i = 1; i < events.length; i++) {
+      const _events = events[i];
+      _events.forEach((event) => {
+        if (!uniqueEventKeys.includes(event.transactionHash)) {
+          missingEvents.push(event);
+        }
+      });
+    }
```

Comment:
A few points on this loop:
- It looks like this is specifically looking for events that providers 2, 3, 4... have that are not included in provider 1. Do we care about the case where 1 has an event, but providers 2, 3, 4... _don't_? The former is the problem we've seen before, but do we want to cover the other form of "disagreement"?
- The way the loop is designed, if providers 2, 3, 4 all have the same event that 1 is missing, we would have that event show up 3+ times in our missingEvents array, right? Wouldn't we only want that to show up once since it's the same event?

Reply:
I've updated this core logic now. This helper method now returns two sets of EventData arrays:
- `missingEvents`: any events found in provider 1 and NOT found in ALL of providers 2 through n
- `matchedEvents`: any events found in provider 1 and also found in ALL of providers 2 through n

There should be no duplicate events anymore, PTAL!